### DISCUSSION TITLE
有关读流程代码的一些疑问

### DISCUSSION THREAD

**Discussion by Lawliet828** [juicedata/juicefs#5390] [General]
### 1

[readAhead函数](https://github.com/juicedata/juicefs/blob/main/pkg/vfs/reader.go#L525)

![image](https://github.com/user-attachments/assets/ab6b1be9-c5eb-4b15-a532-5849983a23fa)

无论是否进入 if r.state == READY 分支，block.len 最终都会被重新设置，要不`block.len = block.end() - r.block.end()`要不`block.len = 0`，这使得 `block.len -= f.r.blockSize / 2` 的操作没有实际效果，所以红框中的代码是无效代码吧。

### 2

[run函数](https://github.com/juicedata/juicefs/blob/main/pkg/vfs/reader.go#L160)

<img width="551" alt="image" src="https://github.com/user-attachments/assets/599def9c-c2fa-4a48-94cc-ec8c972ba4ec" />
<img width="641" alt="image" src="https://github.com/user-attachments/assets/09632b7e-0b5c-49cd-a866-937202ebbc1d" />

162行`f.Lock()`后，163行`defer f.Unlock()`，然后170行`f.Unlock()`，会让人比较迷惑，因为double unlock会引发运行时错误。

实际上162行`f.Lock()`和170行`f.Unlock()`是一对，174行`f.Lock()`和200行`f.Unlock()`是一对，163行的`defer f.Unlock()`应该与208行`f.Lock()`配对使用，所以`defer f.Unlock()`移到208行之后更合理。

**Comment by jiefenghuang** [comment]
1. end()会变化
2. 中间可能退出. 比如165行
